Nodejs/Product/Nodejs/SharedProject/HierarchyIdMap.cs

```diff
@@ -1,25 +1,25 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
+using System.Collections.Concurrent;
 using System.Diagnostics;
 
 namespace Microsoft.VisualStudioTools.Project
 {
     internal sealed class HierarchyIdMap
     {
-        private readonly List<WeakReference<HierarchyNode>> ids = new List<WeakReference<HierarchyNode>>();
-        private readonly Stack<int> freedIds = new Stack<int>();
-
-        private readonly object theLock = new object();
+        private readonly ConcurrentDictionary<int, WeakReference<HierarchyNode>> nodes = new ConcurrentDictionary<int, WeakReference<HierarchyNode>>();
+        private readonly ConcurrentStack<int> freedIds = new ConcurrentStack<int>();
 
         /// <summary>
         /// Must be called from the UI thread
         /// </summary>
         public uint Add(HierarchyNode node)
         {
+            VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread();
```

Comment:
>VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread(); [](start = 12, length = 55)

Do we depend on this or is it just true?  If the latter, consider asserting instead.

Reply:
Seems like we do depend on this.  If it's not true, then we can't just assert that `TryAdd` succeeds.

---
In reply to: [182269818](https://github.com/Microsoft/nodejstools/pull/1933#discussion_r182269818) [](ancestors = 182269818)